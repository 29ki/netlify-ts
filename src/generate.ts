import { pullCollection } from "./collection";
import { appendExport, formatType } from "./output";
import { resolveWidget, transformType } from "./widget";
import { Collection } from "./types";

export const generateTypes = (collections: Collection[]): string => {
  return collections
    .flatMap(pullCollection)
    .map(resolveWidget)
    .reduce(transformType(), [[], []])
    .flat()
    .map(formatType)
    .map(appendExport)
    .join("\n\n")
    .replace(/^/, '/* eslint-disable */\n/* tslint:disable */\n\n')
    .concat("\n");
};
