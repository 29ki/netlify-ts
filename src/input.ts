import yaml from "js-yaml";
import { promises as fs } from "fs";
import path from "path";
import type { Collection } from "./types";

interface YamlInput {
  collections?: Collection[];
}

export const loadConfig = async (filePath: string): Promise<Collection[]> => {
  try {
    if (!path.extname(filePath).match(/\.ya?ml$/)) {
      return Promise.reject("Invalid filetype, must be an yaml file");
    }

    const file = await fs.readFile(path.resolve(process.cwd(), filePath), "utf8");

    const data = yaml.load(file) as YamlInput;

    if (typeof data !== "object" || !data.collections) {
      return Promise.reject("Failed loading collections from config file");
    }

    return data.collections;
  } catch (error) {
    return Promise.reject(error);
  }
};
