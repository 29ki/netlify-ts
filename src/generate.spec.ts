import { loadConfig } from "./input";
import { generateTypes } from "./generate";

describe("Output testing", () => {
  it("should parse kitchen sink correctly", async () => {
    const collections = await loadConfig("kitchen-sink.yml");

    expect(generateTypes(collections)).toMatchSnapshot();
  });
});
