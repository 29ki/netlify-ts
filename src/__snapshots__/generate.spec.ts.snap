// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output testing should parse kitchen sink correctly 1`] = `
"export interface posts {
  title: string;
  draft: boolean;
  date: string;
  image?: string;
  body: string;
}

export interface faq {
  title: string;
  body: string;
}

export interface general_posts {
  front_limit: string;
  author: string;
  thumb?: string;
}

export interface general {
  site_title: string;
  posts: general_posts;
}

export interface authors_authors {
  name: string;
  description: string;
}

export interface authors {
  authors: authors_authors[];
}

export type kitchenSink_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export type kitchenSink_select_multiple_options = \\"a\\" | \\"b\\" | \\"c\\";

export type kitchenSink_select_numeric_options = 1 | 2 | 3;

export type kitchenSink_object_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export interface kitchenSink_object {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: any;
  select: kitchenSink_object_select_options;
}

export type kitchenSink_list_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export type kitchenSink_list_object_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export type kitchenSink_list_object_list_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export type kitchenSink_list_object_list_object_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export interface kitchenSink_list_object_list_object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: any;
  select: kitchenSink_list_object_list_object_select_options;
}

export interface kitchenSink_list_object_list {
  post: any;
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: any;
  select: kitchenSink_list_object_list_select_options;
  hidden: any;
  object: kitchenSink_list_object_list_object;
}

export interface kitchenSink_list_object {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: any;
  select: kitchenSink_list_object_select_options;
  list: kitchenSink_list_object_list[];
}

export interface kitchenSink_list {
  string: string;
  boolean: boolean;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  image: string;
  file: any;
  select: kitchenSink_list_select_options;
  object: kitchenSink_list_object;
}

export interface kitchenSink_typed_list_type_2_object_nested {
  number: string;
}

export type kitchenSink_typed_list_type_2_object_select_options = \\"a\\" | \\"b\\" | \\"c\\";

export interface kitchenSink_typed_list_type_1_object {
  type: \\"type_1_object\\";
  string: string;
  boolean: boolean;
  text: string;
}

export interface kitchenSink_typed_list_type_2_object {
  type: \\"type_2_object\\";
  number: string;
  nested: kitchenSink_typed_list_type_2_object_nested;
  select: kitchenSink_typed_list_type_2_object_select_options;
  datetime: string;
  markdown: string;
}

export interface kitchenSink_typed_list_type_3_object {
  type: \\"type_3_object\\";
  date: string;
  image: string;
  file: any;
}

export interface kitchenSink_code {
  code: string;
  lang: string;
}

export interface kitchenSink_code_alt {
  alt: string;
  typescript: string;
}

export interface kitchenSink {
  post: any;
  title: string;
  boolean: boolean;
  map: string;
  text: string;
  number: string;
  markdown: string;
  datetime: string;
  date: string;
  color: string;
  colorEditable: string;
  image: string;
  file: any;
  select: kitchenSink_select_options;
  select_multiple: kitchenSink_select_multiple_options[];
  select_numeric: kitchenSink_select_numeric_options;
  hidden: any;
  object: kitchenSink_object;
  list: kitchenSink_list[];
  typed_list: (kitchenSink_typed_list_type_1_object | kitchenSink_typed_list_type_2_object | kitchenSink_typed_list_type_3_object)[];
  code: kitchenSink_code;
  code_alt: kitchenSink_code_alt;
  snippet: string;
}
"
`;
